Sidechain creation version

ptagl opened the pull request:
This PR introduces the concept of "Sidechain version" in the Sidechain Creation Output.

To make the change easier, such version field has not been just added to `CTxScCreationOut` but it's been stored together with the `withdrawalEpochLength`.

So, before these changes, the serialization of a `CTxScCreationOut` was reading/writing 4 bytes (32 bits) for the `withdrawalEpochLength`, now instead the first (most significant byte) of the `withdrawalEpochLength` is used for the sidechain `version`. This is possible because even though the `withdrawalEpochLength` had space for 32 bits integers, the maximum allowed value is 4032 (using only 12 bits). For this reason, we can assume that the sidechain version before this PR was already set to 0.

```
              32 bits | 4 bytes                           8 bits | 1 byte      24 bits | 3 bytes

+-------------------------------------------+            +---------------+---------------------------+
|                                           |            |               |                           |
|                                           |            |               |                           |
|           withdrawalEpochLength           | +--------> |    version    |   withdrawalEpochLength   |
|                                           |            |               |                           |
|                                           |            |               |                           |
+-------------------------------------------+            +---------------+---------------------------+

```


List of changes:
- Added Fork 9 (Sidechain Version) that activates at block height 1100000, 1000000, 450 for Mainnet, Testnet and Regtest (**numbers to be set properly**)
- Added the `version` field in `CTxScCreationOut` and in `ScFixedParameters`
- Added the `sidechainVersion` parameter to the `GetFieldElement` function to select the proper validation logic
- Updated RPC commands `sc_create`, `createrawtransaction`, `getscinfo` to include the `version` argument (mandatory when creating a new sidechain)
- Added unit and regression tests

Additionally, while developing unit and regression tests, some helper classes have been created/extended to make it easier the definition of common scenarios (e.g. crafting a sidechain creation transaction, testing the `AcceptTxToMemoryPool()` function, etc.).

Commit:
Prepared the management of Fork 9 (sidechain version)

Commit:
Added the sidechain version in CTxScCreationOut and in ScFixedParameters

Commit:
Added the sidechain version as a parameter for custom fields validation

Commit:
Sidechain version fork height management

Modified the ContextualCheck() function to enforce that a sidechain creation transaction uses only the allowed version (before and after the Sidechain Version fork). This function is called both in AcceptTxToMemoryPool() and in AcceptBlock().

In order to implement some unit tests for all the scenarios, the test framework has been extended to call AcceptTxToMemoryPool() safely.

Commit:
Added regressions tests for the sidechain version management

While implementing the tests, some fixes have been made to Zend (in particular to RPC commands) to correctly handle the version field.
Additionally, fixed some Python tests that stopped working due to the sidechain version.

Commit:
Added a Python test to check the sidechain version fork point

In order to develop such test, a BlockchainHelper Python class has been developed to hide some complexity.

It needs to be extended in the future to address more scenarios.

Commit:
Set the "version" parameter in "sc_create" RPC command as mandatory

Commit:
Made the "version" argument mandatory in the sc_create() RPC command

Commit:
Made "version" argument mandatory in "createrawtransaction" sc creation

Commit:
Added a test for sc_create() RPC command without version

Commit:
Added "sidechain version" to "getscinfo()" RPC command

Some tests have also been implemented.

Commit:
Used the constant SC_VERSION_FORK_HEIGHT in some Python tests

Commit:
Set sc_version to 0xff by default to detect not initialized variables

Commit:
Fixed tests

Commit:
Extended Python tests to verify the persistence of the sc version info

Commit:
Fixed typo and removed unused code

Commit:
Restored the serialization of a field in CSidechain objects

This was needed to avoid to break compatibility with sidechains created before fork 9.

Commit:
Implemented additional UTs for custom fields validation

alsala commented on src/sc/sidechain.h:
```diff
@@ -132,7 +130,17 @@ class CSidechain {
+            The initial "sidechainVersion" variable has then been removed to avoid confusion and inconsistencies and its
+            space in the serialization is held by this dummy/reserved variable.
+        */
+        uint32_t reserved = 0xffffffff;
```
Minor: probably it is best to keep 0 values in order not to have any difference with the value stored so far

ptagl replied:
Fixed with commit [ccdf6c5a5feb54b6353030ee8afec4570281d25d](https://github.com/HorizenOfficial/zen/pull/435/commits/ccdf6c5a5feb54b6353030ee8afec4570281d25d).

i-Alex commented on qa/rpc-tests/sc_create.py:
```diff
@@ -72,13 +72,39 @@ def run_test(self):
         # ---------------------------------------------------------------------------------------
-        # Node 2 try creating a SC with insufficient funds
-        mark_logs("\nNode 2 try creating a SC with insufficient funds", self.nodes, DEBUG_MODE)
+        # Node 2 tries to create a SC without version
```
what if user will try to set `version` > 1?
let's add the test as well.

ptagl replied:
Any `version > 1` will be rejected by the RPC command (and obviously there are the same checks in `AcceptTxToMemoryPool()` and `ConnectBlock()` functions).

Python test extended with commit [0b50c831acf78c8ee7ce4b28a3d9280bba83896b](https://github.com/HorizenOfficial/zen/pull/435/commits/0b50c831acf78c8ee7ce4b28a3d9280bba83896b).

i-Alex commented on src/primitives/transaction.cpp:
```diff
@@ -1069,13 +1072,24 @@ bool CTransaction::ContextualCheck(CValidationState& state, int nHeight, int dos
+            //enforce that any eventual SC creation output is using a valid sidechain version for the current active fork
+            for (const CTxScCreationOut& scCreationOutput : vsc_ccout)
+            {
+                if (scCreationOutput.version > maxSidechainVersionAllowed)
```
Why we should allow sc version `0` after the `sc_version_fork`?
Why not to allow only the sc with version `1`?

ptagl replied:
We decided to allow the creation of old-style sidechains in order to let users who are prepared to create a sidechain to be able to do that also after the fork point with minimal changes (right now just updating the call to `sc_create` RPC command).

i-Alex commented on src/primitives/transaction.h:
```diff
@@ -585,6 +585,20 @@ friend class CTransaction;
+    /**
+     * @brief  Version of the sidechain.
+     * The sidechain version has been introduced with hard fork 9 to increase the flexibility of the sidechain mechanism.
+     * Since the Sidechain Creation Output was at the beginning designed without the version, a single byty for such field
```
`byty` -> `byte`

ptagl replied:
Fixed with commit [4481d65fe92c2936e85e975017fdc573b40257e8](https://github.com/HorizenOfficial/zen/pull/435/commits/4481d65fe92c2936e85e975017fdc573b40257e8).

i-Alex commented on src/sc/sidechaintypes.h:
```diff
@@ -509,6 +511,7 @@ struct ScFixedParameters
     ADD_SERIALIZE_METHODS;
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(version);
```
I think we can't add `READWRITE(version)` for free for `ScFixedParameters` struct.
`ScFixedParameters` is used as a part of `CSidechain` class serialization.
The node without these changes will not be able to recover the old `CSidechain` records from the database after the update to the recent code base.
We should have the same serialization approach as inside Transaction.

ptagl replied:
Fixed with commit [e888104ef55902c02162b2a31abb7b6cfc679b13](https://github.com/HorizenOfficial/zen/pull/435/commits/e888104ef55902c02162b2a31abb7b6cfc679b13).

i-Alex commented on src/gtest/test_mempool.cpp:
```diff
@@ -351,6 +351,65 @@ TEST(Mempool, SproutNegativeVersionTx) {
+
+    // Check that a Sidechain Creation transaction with version 1 is accepted
+    state = CValidationState();
+    EXPECT_EQ(MempoolReturnValue::VALID, testManager.TestAcceptTxToMemoryPool(state, mtx_v1));
```
Check the behavior of the `CTxMemPool::removeStaleTransactions`. We should remove Txs with sc_creation version `1` in case of the rollback before the fork, for example: `testManager.ExtendChainActiveToHeight(sidechainVersionForkHeight - 1);`
Note: maybe `removeStaleTransactions` is not the best method where to put `sc cr version` check, but for sure we need it.

ptagl replied:
After double checking the code, we noticed that we didn't include a similar check for the Zendoo fork point and there must be a reason for that.

In the end we recalled that the we don't ever have to include fork point checks in the `CTxMemPool::removeStaleTransactions` function since it is called only when disconnecting a tip after the generation of a blockchain fork. In this case, we are not only disconnecting blocks, but also reconnecting some others and we can be sure that the final height will be higher than the previous one (otherwise the fork couldn't have led to the disconnection of the tip) and thus we don't need to validate the transactions against the fork point activation.

i-Alex commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -1030,580 +1058,593 @@ TEST(CctpLibrary, BitVectorCertificateFieldFullBzip2)
-    {
-        wrappedFieldPtr sptrScId = CFieldElement(ccout.GetScId()).GetFieldElement();
-        field_t* scid_fe = sptrScId.get();
+        int custom_fields_len = cert.vFieldElementCertificateField.size() + cert.vBitVectorCertificateField.size(); 
```
We need to discuss the cert custom field usage in the commitment tree before and after the fork.

ptagl replied:
In order to let forgers know the version of other sidechains, we added a new websocket API to retrieve such information given a list of sidechain IDs as input (see commit [967bdd0bdf8aa6d7998c718b8ad47659795971ac](https://github.com/HorizenOfficial/zen/pull/435/commits/967bdd0bdf8aa6d7998c718b8ad47659795971ac)).

alsala commented on src/wallet/rpcwallet.cpp:
```diff
@@ -821,6 +822,20 @@ UniValue sc_create(const UniValue& params, bool fHelp)
+    {
+        sidechainVersion = find_value(inputObject, "version").get_int();
+
+        if (sidechainVersion < 0 || sidechainVersion > ForkManager::getInstance().getHighestFork()->getMaxSidechainVersion())
```
We could also check the specific fork version using `chainActive.Height()`

ptagl replied:
Fixed with commit [b35e836f000e35d3356fecb882fb6dc17b24ae96](https://github.com/HorizenOfficial/zen/pull/435/commits/b35e836f000e35d3356fecb882fb6dc17b24ae96).

Commit:
Fixed typo

Commit:
Fixed typo

Commit:
Reverted "reserved" CSidechain field to 0 instead of 0xffffffff

Commit:
Fixed a breaking change in the ScFixedParameters serialization

Commit:
Fixed a wrong check for the sidechain version fork point in sc_create

Commit:
Added a test for the creation of sidechains with totally invalid version

Commit:
Fixed an issue with the Python blockchain helper class

Commit:
Added GetSidechainVersions websocket API

Commit:
Added Python tests for GetSidechainVersions websocket API

albertog78 commented on src/primitives/transaction.h:
```diff
@@ -610,7 +625,27 @@ friend class CTransaction;
+        }
+        else
+        {
+            int withdrawalEpochLengthAndVersion = (version << 24) | withdrawalEpochLength;
```
Would be useful to consider withdrawalEpochLength & 0x00FFFFFF  and version & 0xFF or assert they are in valid ranges?

ptagl replied:
I think that we can skip the check for the `version` field due to its `uint8_t` datatype (so it will never be greater than 0xFF). It might instead be useful for the `withdrawalEpochLength` since it could overwrite the `version` during the OR bitwise operation, but I'm not sure if it's possible also in practice since the `withdrawalEpochLength` is limited by the chain parameters (see `nScMaxWithdrawalEpochLength` in `chainparams.cpp`).

Do you have any specific scenario in mind that we should take care of?

albertog78 replied:
Agree about version. For what regards withdrawalEpochLength the mask (or assert) would be useful to detect / prevent buggy clients generating messy txs (eg. when tx is not yet in a block / mempool then not validated yet by consensus rules).

albertog78 commented on src/sc/sidechaintypes.h:
```diff
@@ -509,7 +511,25 @@ struct ScFixedParameters
+        }
+        else
+        {
+            int withdrawalEpochLengthAndVersion = (version << 24) | withdrawalEpochLength;
```
see comment in CTxScCreationOut serialization code

Commit:
Added a check when serializing "version" and "withdrawalEpochLength"

Commit:
Added version of sidechains that sent a certificate in getscgenesisinfo

Commit:
Fixed a bug in getscgenesisinfo RPC command

Commit:
Added a Python test for sc_getscgenesisinfo RPC command

Additionally, improved the Python BlockchainHelper to easily send certificates.

Commit:
Minor fix for FieldElementCertificateFieldConfig toString()

Commit:
Fixed execution permissions for sc_getscgenesisinfo.py test

Commit:
Struct sSidechainVersion_tag renamed as ScVersionInfo

Commit:
Added some utility test functions

cronicc commented on src/zen/forks/fork9_sidechainversionfork.cpp:
```diff
@@ -0,0 +1,14 @@
+    // TODO: set proper fork height values.
+    setHeightMap({{CBaseChainParams::Network::MAIN,1100000},
+                  {CBaseChainParams::Network::REGTEST,450},
+                  {CBaseChainParams::Network::TESTNET,1000000}});
```
Set higher temporary activation heights, these block heights have already passed on testnet/mainnet.

ptagl replied:
Fixed with commit [190d49622f7e7e70a9d60a74426dc82c813bbb20](https://github.com/HorizenOfficial/zen/pull/435/commits/190d49622f7e7e70a9d60a74426dc82c813bbb20).

cronicc commented on src/gtest/test_forkmanager.cpp:
```diff
@@ -286,3 +287,20 @@ TEST(ForkManager, SidechainForkMainnet) {
+    SelectParams(CBaseChainParams::MAIN);
+
+    // TODO: set proper fork height value.
+    int sidechainVersionForkHeight = 1100000;
```
Set higher temporary activation height. Extend test with testnet and regtest coverage.

ptagl replied:
Fixed with commit [190d49622f7e7e70a9d60a74426dc82c813bbb20](https://github.com/HorizenOfficial/zen/pull/435/commits/190d49622f7e7e70a9d60a74426dc82c813bbb20).

Commit:
Increased fork point activation height for sidechain version

Added unit tests for the activation also for testnet and regtest.

cronicc committed:
Merge pull request #437 from HorizenOfficial/sc_version_in_genesis_info
